File: sudoku.go
package main

import (
	"fmt"
    "bufio"
    "os"
    "strings"
    "strconv"
)

func Populate(fileName string) [][]int {
    inputFile, err := os.Open(fileName)
    if err != nil {
        panic(err)
    }
    defer inputFile.Close()
    inputBoard := [][]int{}
    scanner := bufio.NewScanner(inputFile)
    for scanner.Scan() {
        inputLine := []int{}
        for _, element := range(strings.Split(scanner.Text(), ",")) {
            num, err := strconv.Atoi(element)
            if err != nil {
                panic(err)
            }
            inputLine = append(inputLine, num)
        }
        inputBoard = append(inputBoard, inputLine)
    }
	return inputBoard
}

func IsNotDup(num int, seen []bool) bool {
	if num != 0 {
		if seen[num - 1] {
			return false
		}
		seen[num - 1] = true
	}
	return true
}

func Clear(seen []bool) {
    for i := range seen {
        seen[i] = false;
    }
}

func Invalid(board [][]int) bool {
    row := []bool {false, false, false, false, false, false, false, false, false};
    column := []bool {false, false, false, false, false, false, false, false, false};
    box := []bool {false, false, false, false, false, false, false, false, false};
	for i := range board {
		Clear(row)
		Clear(column)
		Clear(box)
		for j := range board {
			result := IsNotDup(board[i][j], row) && IsNotDup(board[j][i], column) && IsNotDup(board[3*(i/3)+j/3][3*(i%3)+j%3], box)
			if !result {
				return true
			}
		}
	}
	return false
}

func AppendAll(a [][][]int, b [][][]int) [][][]int {
	if len(b) == 0 {
		return a
	}
	c := make([][][]int, len(a)+len(b))
	for i, board := range a {
		c[i] = board
	}
	for i, board := range b {
		c[i+len(a)] = board
	}
	return c
}

func Copy(in [][]int) [][]int {
	out := make([][]int, len(in))
	for i, sub := range in {
		out[i] = make([]int, len(sub))
		for j, num := range sub {
			out[i][j] = num
		}
	}
	return out
}

func GuessAndCheck(board [][]int, count *int) [][][]int {
	*count += 1
	if *count%10000 == 0 {
		fmt.Println(*count)
	}
	if Invalid(board) {
		return [][][]int{}
	}
	validGuesses := [][][]int{}
	for i := range board {
		for j := range board {
			if board[i][j] == 0 {
				for guess := 1; guess < 10; guess++ {
					board[i][j] = guess
					validGuesses = AppendAll(validGuesses, GuessAndCheck(board, count))
				}
				board[i][j] = 0
				return validGuesses
			}
		}
	}
	//Nothing has been returned, hence the board must be complete and valid
	return [][][]int{Copy(board)}
}

func PrettyPrint(board [][]int) {
	for _, row := range board {
		fmt.Println(row)
	}
}

func main() {
	board := Populate(os.Args[1])
	PrettyPrint(board)
	fmt.Println("")
	count := 0
    fmt.Println("all is well:")
	solutions := GuessAndCheck(board, &count)
	fmt.Println("computed", len(solutions), "solution(s), after", count, "guesses")
	//for _, solution := range solutions {
	//	fmt.Println("")
	//	PrettyPrint(solution)
	//}
}

File: Sudoku.java
import java.util.ArrayList;
import java.util.List;
import java.io.*;

public class Sudoku {
    public static void print(int[] array){
        for(int element: array){
            System.out.print(element);
            System.out.print(" ");
        }
    }

    public static void print(int[][] array) {
        for(int[] element: array){
            print(element);
            System.out.print("\n");
        }
    }

    private static int[][] populate(String fileName) throws Exception {
        String line;
        BufferedReader fileReader = new BufferedReader(new FileReader(fileName));
        int[][] inputBoard = new int[9][9];
        int i = 0;
        while((line=fileReader.readLine())!=null){
            String[] lineElements = line.split(",");
            for(int j = 0; j < lineElements.length; j++) {
                inputBoard[i][j] = Integer.parseInt(lineElements[j]);
            }
            i++;
        }
        return inputBoard;
    }

    private static void clear(boolean[] seen){
        for(int i=0;i<9;i++){
            seen[i] = false;
        }
    }

    private static boolean isNotDup(int num, boolean[] seen){
        if(num != 0){
            if(seen[num - 1]){
                return false;
            }
            seen[num - 1] = true;
        }
        return true;
    }

    private static boolean invalid(int[][] board){
        boolean row[] = new boolean[9];
        boolean column[] = new boolean[9];
        boolean box[] = new boolean[9];
        for(int i=0; i<board.length; i++) {
            clear(row);
            clear(column);
            clear(box);
            for(int j=0; j<board[i].length; j++) {
                boolean result = isNotDup(board[i][j], row) && isNotDup(board[j][i], column) && isNotDup(board[3*(i/3)+j/3][3*(i%3)+j%3], box);
                if(!result){
                    return true;
                }
            }
        }
        return false;
    }

    private static List<int[][]> guessAndCheck(int[][] board, int[] count) {
        count[0] += 1;
        if(count[0]%10000 == 0) {
            System.out.println(count[0]);
        }
        List<int[][]> validGuesses = new ArrayList<int[][]> ();
        if(invalid(board)){
            return validGuesses;
        }
        for(int i=0; i<board.length; i++) {
            for(int j=0; j<board[i].length; j++) {
                if(board[i][j] == 0){
                    for(int guess=1; guess<10; guess++){
                        board[i][j] = guess;
                        validGuesses.addAll(guessAndCheck(board, count));
                    }
                    board[i][j] = 0;
                    return validGuesses;
                }
            }
        }
        ////Nothing has been returned, hence the board must be complete and valid
        validGuesses.add(board);
        return validGuesses;
    }

    public static void main(String args[]) {
        int[][] board;
        try {
            board = populate(args[0]);
        }
        catch(Exception exc) {
            System.out.println(exc);
            return;
        }
        System.out.println("Solving:\n");
        print(board);

        int count[] = new int[] {0};
        List<int[][]> solutions = guessAndCheck(board, count);
        System.out.println("computed " + solutions.size() + " solution(s), after " + count[0] + " guesses:\n\n");
    }
}


